aws/service.ts

```diff
@@ -381,73 +381,89 @@ function computeContainerDefinitions(
-                //
-                // See https://github.com/terraform-providers/terraform-provider-aws/issues/3401.
-                hostPort: p.targetPort || p.port,
-            }));
-
-            return pulumi.all([imageOptions, container, logGroup.id])
-                         .apply(([imageOptions, container, logGroupId]) => {
-                const keyValuePairs: { name: string, value: string }[] = [];
-                for (const key of Object.keys(imageOptions.environment)) {
-                    keyValuePairs.push({ name: key, value: imageOptions.environment[key] });
-                }
+    return pulumi.all(containerDefinitions);
+}
 
-                const containerDefinition: aws.ecs.ContainerDefinition = {
-                    name: containerName,
-                    image: imageOptions.image,
-                    command: container.command,
-                    cpu: container.cpu,
-                    memory: container.memory,
-                    memoryReservation: container.memoryReservation,
-                    portMappings: portMappings,
-                    environment: keyValuePairs,
-                    mountPoints: (container.volumes || []).map(v => ({
-                        containerPath: v.containerPath,
-                        sourceVolume: (v.sourceVolume as Volume).getVolumeName(),
-                    })),
-                    logConfiguration: {
-                        logDriver: "awslogs",
-                        options: {
-                            "awslogs-group": logGroupId,
-                            "awslogs-region": aws.config.requireRegion(),
-                            "awslogs-stream-prefix": containerName,
-                        },
-                    },
-                    dockerLabels: container.dockerLabels,
-                };
-                return containerDefinition;
-            });
-        });
+function computeContainerDefinition(
+        parent: pulumi.Resource,
+        containerName: string,
+        container: cloud.Container,
+        imageName: string,
+        ports: ExposedPorts | undefined,
+        logGroup: aws.cloudwatch.LogGroup): pulumi.Output<aws.ecs.ContainerDefinition> {
 
-    return pulumi.all(containerDefinitions);
+    if (!imageName) {
+        throw new Error("[getImageName] should have always produced an image name.");
+    }
+
+    let repository: aws.ecr.Repository | undefined;
+    if (container.build) {
+        // Create the repository.  Note that we must do this in the current turn, before we hit any
+        // awaits. The reason is subtle; however, if we do not, we end up with a circular reference
+        // between the TaskDefinition that depends on this repository and the repository waiting for
+        // the TaskDefinition, simply because permitting a turn in between lets the TaskDefinition's
+        // registration race ahead of us.
+        repository = getOrCreateRepository(imageName);
```

Comment:
This used to run outside of `apply` but now runs inside, right?  And I believe it may create resources.  Also - I admit I don’t understand the comment above this - but it seems like this change violates what it says must happen.